pep-0xxx.rst

```diff
@@ -0,0 +1,1149 @@
+form that trust stores take.
+
+This problem is the same as the one that the Certificate and PrivateKey types
+need to solve. For this reason, we use the exact same model, by creating an
+opaque type that can encapsulate the various means that TLS backends may open
+a trust store.
+
+A given TLS implementation is not required to implement all of the
+constructors. However, it is strongly recommended that a given TLS
+implementation provide the ``system`` constructor if at all possible, as this
+is the most common validation trust store that is used. Concrete
+implementations may also add their own constructors.
+
+Concrete implementations should aim to have TrustStore objects be hashable if
+at all possible. This will help ensure that TLSConfiguration objects used with
+an individual concrete implementation are also hashable.
+
+    class TrustStore(metaclass=ABCMeta):
+        @abstractclassmethod
+        def system(cls) -> TrustStore:
+            """
+            Returns a TrustStore object that represents the system trust
+            database.
+            """
+
+        @abstractclassmethod
+        def from_pem_file(cls, path: Union[pathlib.Path, bytes, str]) -> TrustStore:
+            """
+            Initializes a trust store from a single file full of PEMs.
+            """
+
+
+Changes to the Standard Library
+===============================
+
+The portions of the standard library that interact with TLS should be revised
+to use these ABCs. This will allow them to function with other TLS backends.
+This includes the following modules:
+
+- asyncio
+- ftplib
+- http.client
+- imaplib
+- nntplib
+- poplib
+- smtplib
+
+
+Future
+======
+
+Major future TLS features may require revisions of these ABCs. These revisions
+should be made cautiously: many backends may not be able to move forward
+swiftly, and will be invalidated by changes in these ABCs. This is acceptable,
+but wherever possible features that are specific to individual implementations
+should not be added to the ABCs. The ABCs should restrict themselves to
+high-level descriptions of IETF-specified features.
+
+
+ToDo
+====
+
+* Consider adding a new parameter (``valid_subjects``?) to ``wrap_socket`` and
+  ``wrap_buffers`` that specifies in a *typed* manner what kind of entries in
+  the SAN field are acceptable. This would break the union between SNI and
+  cert validation, which may be a good thing (you can't SNI an IP address, but
+  you can validate a cert with one if you want).
+* It's annoying that there's no type signature for fileobj. Do I really have to
+  define one as part of this PEP? Otherwise, how do I define the types of the
+  arguments to ``wrap_buffers``?
+* Do we need ways to control hostname validation?
+* Do we need to support getpeercert? Should we always return DER instead of the
+  weird semi-structured thing?
+* How do we load certs from locations on disk? What about HSMs?
```

Comment:
Looks like you have an answer for locations on disk.

For HSMs you'll commonly see a key reference loaded in PKCS11 by passing a set of properties that uniquely identify the key. I'm not sure I see an advantage to trying to define a `from_opaque_data` constructor on the abstract base class though. Someone wanting to support keys stored in HSMs would simply implement it all themselves and then register their concrete class as conforming to the `PrivateKey` ABC.

Since `PrivateKey` has no required behavior in the current draft the only way an HSM `PrivateKey` could work with existing concrete implementations of this TLS PEP would be for the developer to look at the implementation details and conform to that unofficial interface. If the idea is that each TLS implementation will provide its own more extensive interfaces for `PrivateKey` that might be okay, otherwise we may want to consider the set of operations required.